### PR TITLE
Add list source reference order validation

### PR DESCRIPTION
**Pull request by petechd**
### What is the context of this PR?
Adds validation of sources referencing lists, no longer list source for a given list can be referenced in a section that precedes the section in which the list collector itself for that list is set.

### How to review
New invalid schema added and new unit test that checks the error message.

### Checklist
* [ ] eq-translations updated to support any new schema keys which need translation

### CODE REVIEW

File: tests/test_questionnaire_validator.py [hunk 1]
```diff
@@ -406,3 +406,21 @@ def test_invalid_calculated_or_grand_calculated_summary_id_in_value_source():
     validator.validate()
 
     assert validator.errors == expected_errors
+
+
+def test_list_as_source_referenced_before_created():
```
Comment (berroar) [inline tests/test_questionnaire_validator.py:411]: As discussed, will need to cover the case when the list source is before the collector within the same section also

Comment (petechd) [reply tests/test_questionnaire_validator.py:411]: Thanks, all seems to be added, there might be some very edge cases here and there but I think all is checked now.

File: app/validators/questionnaire_validator.py [hunk 2]
```diff
@@ -134,3 +139,90 @@ def validate_introduction_block(self):
...
+                        # Section level "enabled" rule that can use list source,
+                        # check: common_definitions.json#/section_enabled
+                        self.add_error(
+                            error_messages.LIST_REFERENCED_BEFORE_ADDED.format(
+                                list_name=list_identifier
+                            ),
+                            section_name=section["id"],
+                        )
+
+    def populate_referenced_lists(self):
```
Comment (berroar) [inline app/validators/questionnaire_validator.py:180]: Since this is just populating, I would move this to questionnaire_schema and populate the map up front similar to what we the way we do in runner

Comment (petechd) [reply app/validators/questionnaire_validator.py]: I made it questionnaire schema `@property` similar to `answers_with_context`. Thanks.

File: app/error_messages.py [hunk 3]
```diff
@@ -40,3 +40,6 @@
     "Variants have more than one question type for block."
 )
 PREVIEW_WITHOUT_INTRODUCTION_BLOCK = "No introduction block found. Introduction block is mandatory when using the preview questions feature."
+LIST_REFERENCED_BEFORE_ADDED = (
+    "List '{list_name}' referenced as source before it has been created."
```
Comment (liamtoozer) [inline app/error_messages.py:44]: Quick question - could the list ID be passed in as a key/val to the error message instead, to prevent using string formatting?

File: app/validators/questionnaire_schema.py [hunk 4]
```diff
@@ -209,6 +210,44 @@ def answers_with_context(self):
...
+                    "section_index": 0,
+                    "block_index": 0,
+                }
+
+        if blocks := self.list_collectors:
+            for block in blocks:
+                list_id = block["for_list"]
+                if list_id not in self._lists_with_context:
+                    section_id = self.get_section_id_for_block_id(block["id"])
+                    section_index = self.get_section_index_for_section_id(section_id)
```
Comment (liamtoozer) [inline app/validators/questionnaire_schema.py:230]: Minor learning question! Can `section_index = self.section_ids.index(section_id)` be used here so that a function might not be needed? Or is that not a good idea as it might depend on the order?

File: app/validators/questionnaire_schema.py [hunk 5]
```diff
@@ -617,3 +656,26 @@ def get_section_id_for_block(self, block: Mapping) -> str | None:
     def get_section_id_for_block_id(self, block_id: str) -> str | None:
         if block := self.get_block(block_id):
             return self.get_section_id_for_block(block)
+
+    def get_list_collector_block_id_for_add_block(self, block_id) -> dict | None:
```
Comment (liamtoozer) [inline app/validators/questionnaire_schema.py:660]: I couldn't spot this being used, so removing it might be okay?

File: app/validators/questionnaire_schema.py [hunk 6]
```diff
@@ -617,3 +656,26 @@ def get_section_id_for_block(self, block: Mapping) -> str | None:
...
+        for block in self.blocks:
+            if (
+                block["type"] == "ListCollector"
+                and block["add_block"]["id"] == block_id
+            ):
+                return block["id"]
+
+    def get_list_collector_block_id_for_repeating_blocks(self, block_id) -> dict | None:
```
Comment (liamtoozer) [inline app/validators/questionnaire_schema.py:668]: I couldn't spot this being used, so removing it might be okay?

File: app/validators/questionnaire_validator.py [hunk 7]
```diff
@@ -134,3 +139,38 @@ def validate_introduction_block(self):
...
+                    if parent_block:
+                        if (
+                            self.questionnaire_schema.block_ids.index(
+                                parent_block["id"]
+                            )
+                            < lists_with_context[list_identifier]["block_index"]
+                        ):
+                            self.add_error(
+                                error_messages.LIST_REFERENCED_BEFORE_ADDED.format(
+                                    list_name=list_identifier
```
Comment (liamtoozer) [inline app/validators/questionnaire_validator.py:161]: Might not be possible, but could the parent block ID also be output in this error message, just for some extra info to help narrow down the location in the schema?

File: tests/test_questionnaire_validator.py [hunk 8]
```diff
@@ -406,3 +406,74 @@ def test_invalid_calculated_or_grand_calculated_summary_id_in_value_source():
...
+    filename = "schemas/invalid/test_invalid_list_source_reference.json"
+    validator = QuestionnaireValidator(_open_and_load_schema_file(filename))
+
+    expected_errors = [
+        {
+            "message": "List 'household' referenced as source before it has been "
```
Comment (liamtoozer) [inline tests/test_questionnaire_validator.py:417]: V minor/cosmetic - each `message` on a single line?

**Review by liamtoozer — COMMENTED** [review]
Looks great to me, and thanks for guiding me through the changes! Just had some extra questions

File: app/validators/questionnaire_schema.py (hunk 5 above)
Comment (liamtoozer) [reply app/validators/questionnaire_schema.py:660]: Ah, unless this is for the other PR coming up?

File: app/validators/questionnaire_schema.py (hunk 6 above)
Comment (liamtoozer) [reply app/validators/questionnaire_schema.py:668]: Ah, unless this is for the other PR coming up?

File: app/error_messages.py (hunk 3 above)
Comment (petechd) [reply app/error_messages.py:44]: Yeah, we can do it both ways, up to the devs which way we want this.

File: app/validators/questionnaire_schema.py (hunk 5 above)
Comment (petechd) [reply app/validators/questionnaire_schema.py:660]: No, it's not needed after these simplifications I showed you on Monday, not even needed for that next PR coming up.

File: app/validators/questionnaire_schema.py (hunk 6 above)
Comment (petechd) [reply app/validators/questionnaire_schema.py:668]: Again, it's not needed after these simplifications I showed you on Monday, not even needed for that next PR coming up.

File: app/validators/questionnaire_schema.py (hunk 4 above)
Comment (petechd) [reply app/validators/questionnaire_schema.py:230]: This one is a really good one, how did I miss that opportunity there? Will use your suggestion in both places.

File: app/validators/questionnaire_validator.py (hunk 7 above)
Comment (petechd) [reply app/validators/questionnaire_validator.py:161]: Not a problem,  I will add it as a key/value.

Comment (petechd) [reply app/validators/questionnaire_validator.py:161]: We could possibly keep it as "parent block id" since the block id could be sometimes the nested one, like the add block one.

Comment (petechd) [reply app/validators/questionnaire_validator.py]: I've added it as "parent block id" for now.

File: tests/test_questionnaire_validator.py (hunk 8 above)
Comment (petechd) [reply tests/test_questionnaire_validator.py]: They are now are after we removed list name to it's own key/value pair, thanks.

File: app/error_messages.py (hunk 3 above)
Comment (petechd) [reply app/error_messages.py]: I moved it to a key/val and lets see if other devs like it.

File: tests/test_questionnaire_validator.py (hunk 8 above)
Comment (petechd) [reply tests/test_questionnaire_validator.py]: We can pass the actual error message constant here now, import it, but only when they list id key/val is separated out so not sure if we want to separate it out? @berroar?

Comment (berroar) [reply tests/test_questionnaire_validator.py]: I  think splitting it out is fine (e.g. the current one) because at the end of the day, it is a case of clearly presenting the info to an author, so looks good to me 👍

Comment (petechd) [reply tests/test_questionnaire_validator.py]: Ok, added the changes now, thanks.

File: app/validators/questionnaire_schema.py [hunk 9]
```diff
@@ -209,6 +210,44 @@ def answers_with_context(self):
         self._answers_with_context = answers_dict
         return self._answers_with_context
 
+    @property
+    def lists_with_context(self):
+        if self._lists_with_context:
+            return self._lists_with_context
```
Comment (berroar) [inline app/validators/questionnaire_schema.py:216]: Do we need this if/return, I don't think we would ever hit this at the moment?

File: app/error_messages.py [hunk 10]
```diff
@@ -40,3 +40,4 @@
     "Variants have more than one question type for block."
 )
 PREVIEW_WITHOUT_INTRODUCTION_BLOCK = "No introduction block found. Introduction block is mandatory when using the preview questions feature."
+LIST_REFERENCED_BEFORE_ADDED = "List referenced as source before it has been created."
```
Comment (berroar) [inline app/error_messages.py:43]: Really minor but would make it consistent and use created rather than added as the var name too.

File: app/validators/questionnaire_schema.py (hunk 9 above)
Comment (petechd) [reply app/validators/questionnaire_schema.py:216]: I went with the same approach we use in answers_with_context, since this is a property, we populate it only the initial time.

Comment (petechd) [reply app/validators/questionnaire_schema.py]: I've added the change now after our discussion.

**Review by Yuyuutsu — APPROVED** [review]
Validation works as expected, LGTM !

**Review by berroar — APPROVED** [review]

**Review by liamtoozer — APPROVED** [review]
